### PR TITLE
Add XGBoost

### PR DESCRIPTION
**Pull request by gabrieltseng**
@tommylees112 , as discussed.

This only really works on a big computer (haven't tested on the full dataset yet - we should do that on the ERA5 scale data, since it will be quite a bit smaller).

### CODE REVIEW

File: src/models/base.py [hunk 1]
```diff
@@ -155,3 +157,35 @@ def evaluate(self, save_results: bool = True,
                 ).set_index(['lat', 'lon', 'time']).to_xarray()
 
                 preds_xr.to_netcdf(self.model_dir / f'preds_{key}.nc')
+
+    def _concatenate_data(self, x: Union[Tuple[Optional[np.ndarray], ...],
+                                         TrainData]) -> np.ndarray:
+
```
Comment (tommylees112) [inline src/models/base.py]: can we add a docstring?

```
        """For loading ALL data from dataloaders into memory"""
```

File: src/models/gbdt.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+from .base import ModelBase
+from .data import DataLoader, train_val_mask, TrainData
+
+xgb = None
+
+
+class GBDT(ModelBase):
+
+    """This only really works if you have a lot of memory
+    """
```
Comment (tommylees112) [inline src/models/gbdt.py]:
```
    (all data has to be loaded into memory for the XGBRegressor)

```

Comment (tommylees112) [reply src/models/gbdt.py]: just so it's clear why a user might need alot of memory 

(user being future us ... ;) )

File: src/models/gbdt.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            ignore_vars=self.ignore_vars, monthly_aggs=self.include_monthly_aggs,
+            static=self.include_static)
+
+        preds_dict: Dict[str, np.ndarray] = {}
+        test_arrays_dict: Dict[str, Dict[str, np.ndarray]] = {}
+
+        assert self.model is not None, 'Model must be trained!'
+
+        for dict in test_arrays_loader:
+            for key, val in dict.items():
```
[the remark this answers is not in the document]
Comment (tommylees112) [reply src/models/gbdt.py]: scrap this i think your way of doing it is neater

File: src/models/gbdt.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        for x, y in train_dataloader:
+            input_train_x.append(self._concatenate_data(x))
+            input_train_y.append(y)
+
+        input_train_x_np = np.concatenate(input_train_x, axis=0)
+        input_train_y_np = np.concatenate(input_train_y)
+
+        fit_inputs = {'X': input_train_x_np, 'y': input_train_y_np}
+
+        if early_stopping is not None:
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/models/gbdt.py]: So this is an argument passed to the function. The names are a little confusing, but:

`early_stopping`: patience (number of epochs to wait without improvement). If its `None`, then early stopping isn't used.
`self.early_stopping` (i.e. the model attribute): whether or not to use early stopping

File: src/models/gbdt.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                 include_yearly_aggs: bool = True,
+                 surrounding_pixels: Optional[int] = None,
+                 ignore_vars: Optional[List[str]] = None,
+                 include_static: bool = True) -> None:
+        super().__init__(data_folder, batch_size, experiment, pred_months,
+                         include_pred_month, include_latlons, include_monthly_aggs,
+                         include_yearly_aggs, surrounding_pixels, ignore_vars,
+                         include_static)
+
+        self.early_stopping = False
```
[the remark this answers is not in the document]
Comment (gabrieltseng) [reply src/models/gbdt.py]: see the comment above

File: src/models/base.py (hunk 1 above)
Comment (gabrieltseng) [reply src/models/base.py]: Added!

File: src/models/gbdt.py (hunk 2 above)
Comment (gabrieltseng) [reply src/models/gbdt.py]: Added !

**Pull request merged**